src/nvim/fileio.c

```diff
@@ -2079,6 +2094,658 @@ char *new_file_message(void)
+        ml_timestamp(buf);
+        if (append) {
+          buf->b_flags &= ~BF_NEW;
+        } else {
+          buf->b_flags &= ~BF_WRITE_MASK;
+        }
+      }
+      if (reset_changed && buf->b_changed && !append
+          && (overwriting || vim_strchr(p_cpo, CPO_PLUS) != NULL)) {
+        // Buffer still changed, the autocommands didn't work properly.
+        return FAIL;
+      }
+      return OK;
+    }
+    if (!aborting()) {
+      emsg(_("E203: Autocommands deleted or unloaded buffer to be written"));
+    }
+    return FAIL;
+  }
+
+  // The autocommands may have changed the number of lines in the file.
+  // When writing the whole file, adjust the end.
+  // When writing part of the file, assume that the autocommands only
+  // changed the number of lines that are to be written (tricky!).
+  if (buf->b_ml.ml_line_count != old_line_count) {
+    if (whole) {                                              // write all
+      *endp = buf->b_ml.ml_line_count;
+    } else if (buf->b_ml.ml_line_count > old_line_count) {           // more lines
+      *endp += buf->b_ml.ml_line_count - old_line_count;
+    } else {                                                    // less lines
+      *endp -= old_line_count - buf->b_ml.ml_line_count;
+      if (*endp < start) {
+        no_wait_return--;
+        msg_scroll = msg_save;
+        emsg(_("E204: Autocommand changed number of lines in unexpected way"));
+        return FAIL;
+      }
+    }
+  }
+
+  // The autocommands may have changed the name of the buffer, which may
+  // be kept in fname, ffname and sfname.
+  if (buf_ffname) {
+    *ffnamep = buf->b_ffname;
+  }
+  if (buf_sfname) {
+    *sfnamep = buf->b_sfname;
+  }
+  if (buf_fname_f) {
+    *fnamep = buf->b_ffname;
+  }
+  if (buf_fname_s) {
+    *fnamep = buf->b_sfname;
+  }
+  return NOTDONE;
+}
+
+static void buf_write_do_post_autocmds(buf_T *buf, char *fname, exarg_T *eap, bool append,
+                                       bool filtering, bool reset_changed, bool whole)
+{
+  aco_save_T aco;
+
+  curbuf->b_no_eol_lnum = 0;      // in case it was set by the previous read
+
+  // Apply POST autocommands.
+  // Careful: The autocommands may call buf_write() recursively!
+  aucmd_prepbuf(&aco, buf);
+
+  if (append) {
+    apply_autocmds_exarg(EVENT_FILEAPPENDPOST, fname, fname,
+                         false, curbuf, eap);
+  } else if (filtering) {
+    apply_autocmds_exarg(EVENT_FILTERWRITEPOST, NULL, fname,
+                         false, curbuf, eap);
+  } else if (reset_changed && whole) {
+    apply_autocmds_exarg(EVENT_BUFWRITEPOST, fname, fname,
+                         false, curbuf, eap);
+  } else {
+    apply_autocmds_exarg(EVENT_FILEWRITEPOST, fname, fname,
+                         false, curbuf, eap);
+  }
+
+  // restore curwin/curbuf and a few other things
+  aucmd_restbuf(&aco);
+}
+
+static void set_err_num(Error_T *e, const char *num, const char *msg)
```

Comment:
These can be written as plain assignments i e `err = err_num(num, msg)` at the call site by returning the struct, which makes it immediately clear we are overwriting the entire struct (even if a subset of fields are specificed)

Reply:
Done